Add dune-project generation to dune init

shonfeder opened the pull request:
Closes  #4367

This is still in a broken state, but I'll work it in the evenings this week in the hopes of getting it in by next week.

In the meanwhile, there is at lest one point where I could use input on those with dune expertise, so I'm staging the rough WIP for now to solicit advice at key points.

shonfeder committed:
Document cram stanza in dune-project docs

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Add encode function for Dune_project.t

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Update to latest

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Add support fore subst_config field

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Improve comments

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Add dune-project creation to dune init proj

NOTE: This commit is still buggy and wrong

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder commented on src/dune_engine/dune_project.ml:
```diff
@@ -551,7 +568,96 @@ module Toggle = struct
+     } ->
+  let open Dune_lang.Encoder in
+  let lang = Lang.get_exn "dune" in
+  let extension_args : Dune_lang.t list =
+    let _ = extension_args in
+    (* TODO *)
+    []
```
iiuc, proper encoding of a `dune-project` file will need to extract data from this `Univ_map.t` to synthesize the `using` stanza, but to do that I think we'll need to provide a way of converting a `Univ_map.t` into a less opaque structure (like a list of tuples?). I don't know if the right way to do this would go going through `to_dyn: t -> Dyn.t` (I can't figure out what `Dyn.t` is for, as [the interface is undocumented](https://github.com/ocaml/dune/tree/4ca86d409ef81598750defc3f782e4c90d69c475/otherlibs/stdune-unstable/dyn.mli) ) or to write a new function for the `Univ_map` type. @jeremiedimino: it looks like you've added some functions to the `Univ_map` last year. Any thoughts on this?

Btw, I haven't been able find any docs on this `using` stanza in the dune-project files : a pointer here would be helpful.

jeremiedimino replied:
`Dyn` stands for "dynamic values". The `Dyn.t` type is a representation of OCaml values such that they can be processed without knowing their type. Converting a `Univ_map.t` to a `(string * Dyn.t) list` seems fine, you can add the following function to the `Dyn` module for that:


```ocaml

let to_dyns (t : t) =
  Int.Map.values t
  |> List.map ~f:(fun (Binding.T (key, v)) ->
         let (module K) = key in
         (K.name, K.to_dyn v)))
```

shonfeder replied:
Thanks!

shonfeder replied:
So added that function, but then I also decided I just don't want to fuss with figuring out how to reconstruct this data after it's been split into three different collections, each of which uses different kinds of abstractions that obscure the content. 

I'd welcome additions to this PR to add the proper encoding for this data, and I'd be happy to remove the `to_dyns` function added, but I'm not going to write the encoding of for this data myself, since it is totally orthogonal  for the feature I'm responsible for adding here.

shonfeder committed:
Document purpose of Dyn module

Thanks to @jeremiedimino
see https://github.com/ocaml/dune/pull/4881#discussion_r699204099

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Add Univ_map.to_dyns

Thanks to @jeremiedimino
https://github.com/ocaml/dune/pull/4881#discussion_r699204099

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Don't mess with with extensions stuff

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Add name in package encoding

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Add test for dune-project on init

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Fixes and plumbing

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Fix config-in-workspace test

No longer needs to create its own dune-project file

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder commented on bin/dune_init.ml:
```diff
@@ -230,8 +230,6 @@ module Component = struct
           | Exec
           | Lib
 
-        (* TODO(shonfeder) Add custom templates *)
```
Removed because I don't think we're planning on this any more. But, in any case, I won't be adding it. So if we still want this TODO, I can just remove my name.

jeremiedimino replied:
I'm all for removing uncertain TODOs

shonfeder committed:
Fix formatting

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Clarify comments on omitted fields

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder commented on src/dune_engine/dune_project.ml:
```diff
@@ -551,7 +578,118 @@ module Toggle = struct
+          tho. *)
+     ; extension_args = _
+     ; parsing_context = _
+     ; stanza_parser =
+         _
+         (* The next three fields hold metadata that is about the dune-project
+            file, but not represented in its content *)
```
This is how ocamlformat insisted on formatting this comment. Happy to reformat if given explicit instructions on how you'd prefer it look.

jeremiedimino replied:
The formatting looks good to me. But in general, this is the deal with ocamlformat: the overall benefit is good enough that we accept the occasional formatting we don't like.

shonfeder replied:
I agree it's worth it, and I don't mind the slight blemish :)

shonfeder commented on src/dune_engine/package.ml:
```diff
@@ -357,7 +359,7 @@ module Source_kind = struct
 
     let encode { user; repo; kind } =
       let forge = to_string kind in
-      let path = repo ^ "/" ^ user in
+      let path = user ^ "/" ^ repo in
```
It put them in the wrong order in  #4815

shonfeder commented on src/dune_engine/package.ml:
```diff
@@ -431,6 +433,22 @@ module Info = struct
+             ; repo = "reponame"
+             })
+    ; license = Some "LICENSE"
+    ; authors = Some [ "Author Name" ]
+    ; homepage = Some "https://url/to/home/page"
+    ; bug_reports = Some "https://url/to/issue/tracker"
+    ; documentation = Some "https://url/to/documentation"
+    ; maintainers = Some [ "Maintainer Name" ]
```
Cargo uses git to fill in the author name and email. I don't think that's worth the complication of shelling out to git, and the rigidity of tying initialization to a specific VCS. The point here is to give users a clear and easy template they can expand upon and fill in.

jeremiedimino replied:
If the source is `(github ...)`, these fields are automatically inferred. Given that, what about using `None` for these fields?

shonfeder replied:
Yeah, good idea!

See 8b718b133dc6087b75fb12018bdfc4fdec813051

shonfeder commented on src/dune_engine/package.ml:
```diff
@@ -675,6 +694,28 @@ let file ~dir ~name = Path.relative dir (Name.to_string name ^ opam_ext)
+  ; info = Info.empty
+  ; depopts = []
+  ; has_opam_file = false
+  ; tags = [ "topics"; "to describe"; "your"; "project" ]
```
I include the `"to described"` bit to help show users how to do multi-word topics.

jeremiedimino replied:
That seems reasonable

shonfeder commented on test/blackbox-tests/test-cases/dune-init.t/run.t:
```diff
@@ -302,20 +308,106 @@ Can init and build a new executable project
+
   $ dune exec --root test_exec_proj ./test/test_exec_proj.exe
   Entering directory 'test_exec_proj'
-  $ rm -rf ./test_exec_proj
```
This is not needed any more (if it ever was).

shonfeder commented on test/blackbox-tests/test-cases/dune-init.t/run.t:
```diff
@@ -358,3 +532,8 @@ Can init and build a project using Esy
   dune
   test_esy_proj.ml
+
+And the dune-project file does NOT specify generation of an opam file:
+
+  $ cat test_esy_proj/dune-project | grep "generate_opam_files"
+  (generate_opam_files false)
```
Not sure what other properties should hold of dune-project files in the case of esy projects.

shonfeder reviewed (commented):
Some pointers to help during review.

shonfeder committed:
Fix comments

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
And entry to changelog

Resolving merge conflict
Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Update quickstart to suggest dune init proj

Signed-off-by: Shon Feder <shon.feder@gmail.com>

jeremiedimino commented:
What is the relation between `dune init proj --kind lib` and `dune init lib`?

shonfeder commented:
> What is the relation between dune init proj --kind lib and dune init lib?

`dune init lib` and `dune init exec` and `dune init test` will "initialize" library, executable, or test components, whereas "dune init proj[ect] [--kind=lib/exec]" will initialize a **project**, of the relevant kind.  

You might find this pretty confusing and counterintuitive! I do too. I advocated for a different CLI all along the long way of landing `init`. I made my strongest case against this design [here](https://github.com/ocaml/dune/pull/1448#issuecomment-477124323) and in the following exchange, but finally gave in for the sake of being able to move forward with the getting the basic functionality in place.

shonfeder commented:
Updated to fix link.

jeremiedimino commented:
Thanks for the explanation. I didn't realise that `dune init` could be used to initialise individual components. Who is `dune init` targeted at BTW?

I'm asking a few questions as I haven't followed closely the development of `dune init`, but I do plan to review the PR. Generating a `dune-project` file seems definitely good.

jeremiedimino commented:
@christinerose could you review the changes to the doc/ folder please?

shonfeder commented:
> Who is dune init targeted at BTW

The user I've wanted to serve with project initialization has always been non-experts who just want to get a project going quickly. It's meant to be something comparable to `cargo new` or `stack new` or `npm new` etc. (See, e.g., https://github.com/ocaml/dune/issues/159#issuecomment-408586053).

But I think there were conflicting visions and expectations along the way.  I think the intended target for the sub-component initialization was experienced users who wanted to automate template generation and maybe eventually for editor integrations? I'm not super sure. Relative to my aims, those were just hopes I needed to jump through in order to land the thing I understand  users to be asking for to reduce friction.

If it were up to me, I'd move the project initialization under the `new` subcommand, and make it work mostly just like tools people are familiar with from other tool chains.

christinerose commented:
> @christinerose could you review the changes to the doc/ folder please?
With pleasure!

christinerose commented:
@jeremiedimino — the spelling changes in [4909](https://github.com/ocaml/dune/pull/4909) look good to me. Great catches! Those are the only recent changes I see in the Docs folder. Please let me know if I'm missing others.

shonfeder commented:
Hi, @christinerose! Thank you for reviewing :) 

I am guessing @jeremiedimino is referring to the changes I'm proposing in this PR. This should be just the `rst` files in this changeset: https://github.com/ocaml/dune/pull/4881/files?file-filters%5B%5D=.rst

christinerose commented on doc/dune-files.rst:
```diff
@@ -55,6 +57,18 @@ Sets the version of the project:
 
     (version <version>)
 
+cram
```
Since it's a section header, I've proposed we capitalize these, but I haven't yet heard if those will be accepted or not.

shonfeder replied:
I'll leave this as is for now then, and you and the team can change them all if you end up decided that. I don't think we'd just want this one header capitalized :)

fwiw, I'd find it confusing as a reader if these headers were capitalized, since they correlate to the keywords used in stanzas, and those are not capitalized. But perhaps they should use monospaced text (if that's allowed in headers in rst?).

christinerose replied:
Understood! I got the same feedback from another engineer, so it’s fine to keep them all lowercase.

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
 `examples <https://github.com/ocaml/dune/tree/master/example>`__ for complete
 examples of projects using Dune.
 
+Initializing projects
```
Initializing Project. I suggest capitalizing both words in a header

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+Initializing projects
+=====================
+
+The following subsections illustrate basic usage of the ``dune init proj``
```
The following subsections illustrate basic usage of the ``dune init proj``
subcommand. For more documentation, see :ref:`initializing_components` and the
inline help available from ``dune init --help``.

christinerose replied:
(just a comma after "Fore more documentation")

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+
+.. _initializing-an-executable:
+
+Initializing an executable
```
Initializing an Executable

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+Initializing an executable
+--------------------------
+
+To initialize a project to build an executable program, run the following
```
"To initialize a project that will build an executable program".  ?? 
Just to remove the repetitive "to" ... but it's fine either way, really.

shonfeder replied:
Ah that's better. Thanks :)

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+
+    dune init proj project_name
+
+This will create a directory for the project that includes the following contents:
```
This creates a project directory that includes the following contents: 
...or...
This creates a directory for the project that includes the following contents:

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+* The ``dune`` files in each directory specify the component to be built with
+  the files in that directory. For details on dune files, see :ref:`dune-files`.
+
+Initializing a library
```
Initializing a Library

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+Initializing a library
+--------------------------
+
+To initialize a project for an ocaml library, run the following (replacing
```
capitalize OCaml

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+
+    dune init proj --kind=lib project_name
+
+This will create a directory for the project that includes the following contents:
```
This creates a project directory that includes the following contents:
...or...
This creates a directory for the project that includes the following contents:
(or it's fine the way it is, if you prefer)

shonfeder replied:
I find think those are both better, and the first  one preferable. Thank!

christinerose commented on doc/quick-start.rst:
```diff
@@ -6,10 +6,148 @@ This document gives simple usage examples of Dune. You can also look at
+   dune test
+
+
+All of the sub-components generated are the same as those described in
```
I believe _subcomponents_ doesn't need that hyphen.

christinerose commented on doc/tests.rst:
```diff
@@ -1,3 +1,5 @@
+.. _writing-tests:
+
 *************************
 Writing and running tests
```
I haven't looked over this document yet, but when I do, I'll change these headings to have all words capitalized: 
Writing and Running Tests

shonfeder replied:
:+1:  I didn't change this header as part of this changeset, just added the reference target, but I think that will stay the same.

christinerose commented on doc/tests.rst:
```diff
@@ -469,6 +471,8 @@ which displays nicer looking diffs that the standard ``diff``
 tool. You can change that by passing ``--diff-command CMD`` to
 dune.
```
Capitalize Dune, unless talking about the command `dune`

christinerose reviewed (commented):
Hi @shonfeder!
Thanks for the clarification! I went over those, and they look really good! I had a few comments. (I'm still new to this process with the comments and doing these edits through Git, so I'm not sure if I'm supposed to "Request Change" below or just "Comment"... so I'll just comment for now) :)

shonfeder committed:
Use default for homepage and bug_reports

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Move flags above the package info

Seems to produce a nicer organized default opam file imo.

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Capitalize section headers

Response to review.

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Add comma

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Remove redundant "to"

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Improve phrasing

Suggested by @christinerose

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Fix captialization and phrasing

Suggestions from @christinerose

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder commented:
Thanks for the fixes and improvements, @christinerose! I think I've addressed all comments targeting at things I introduced here, with the exception of capitalizing `cram` in the dune file stanza section, which seems like it should stay consistent in the file.

jeremiedimino commented on src/dune_engine/package.mli:
```diff
@@ -120,10 +120,15 @@ module Info : sig
   val maintainers : t -> string list option
 
+  (** Default package info (e.g., for project initialization ) *)
+  val default : t
```
It feels more like a "sample" than a "default" to me. I'd expect "default" to have all fields set to `None` for instance.

shonfeder replied:
Good call there. I've renamed it to "example".

jeremiedimino reviewed (changes requested):
Looks good to me. Just a nitpick regarding the naming of a value. Kudos for the well written documentation!

shonfeder committed:
Fix typo

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder committed:
Fix name of example package struct

Signed-off-by: Shon Feder <shon.feder@gmail.com>

shonfeder commented:
Typo fixed (thanks @christinerose) and function renamed (thanks @jeremiedimino). Should be ready for another look.

shonfeder commented:
@christinerose Wasn't that typo addressed here? https://github.com/ocaml/dune/pull/4881/commits/256c57f3eeb7de4203adbb643227c18881624d38

Perhaps you are referring to a different error? If so, if you can add your change request to the line you have in mind on the latest set of changes, that would help clarify: https://github.com/ocaml/dune/pull/4881/files

jeremiedimino reviewed (approved):
Looks good to me!

shonfeder commented:
Great. Thanks for the reviews :)

shonfeder commented:
Hmm.. There's a CI failure in the formatting-check stage that's only affecting 4.12 on ubuntu:

```
./dune.exe build @fmt --auto-promote
File "src/dune_engine/package.ml", line 1, characters 0-0:
         git (internal) (exit 1)
(cd _build/default && /usr/bin/git --no-pager diff --no-index --color=always -u src/dune_engine/package.ml src/dune_engine/.formatted/package.ml)
diff --git a/src/dune_engine/package.ml b/src/dune_engine/.formatted/package.ml
index 65ac69a..ea4c0bf 100644
--- a/src/dune_engine/package.ml
+++ b/src/dune_engine/.formatted/package.ml
@@ -444,8 +444,9 @@ module Info = struct
     ; license = Some "LICENSE"
     ; authors = Some [ "Author Name" ]
     ; maintainers = Some [ "Maintainer Name" ]
-    ; documentation = Some "https://url/to/documentation"
-    (* homepage and bug_reports are inferred from the source *)
+    ; documentation =
+        Some "https://url/to/documentation"
+        (* homepage and bug_reports are inferred from the source *)
     ; homepage = None
     ; bug_reports = None
     }
Promoting _build/default/src/dune_engine/.formatted/package.ml to
  src/dune_engine/package.ml.
make: *** [Makefile:92: fmt] Error 1
Error: Process completed with exit code 2.
```

On the face of it, it seems unrelated to any work on this PR, but I don't want to introduce breakage into the mainline. If anyone has an idea of the cause that would be helpful, otherwise, I'll investigate when I get time.

jeremiedimino committed:
formatting

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino commented:
The failure is only with 4.12 because we only test the formatting (and run tests in general) with 4.12. I did `make fmt`, committed and pushed.

shonfeder commented:
Ah, that makes sense. I wasn't aware. Thank you!

christinerose commented:
> @christinerose Wasn't that typo addressed here? [256c57f](https://github.com/ocaml/dune/commit/256c57f3eeb7de4203adbb643227c18881624d38)
> 
> Perhaps you are referring to a different error? If so, if you can add your change request to the line you have in mind on the latest set of changes, that would help clarify: https://github.com/ocaml/dune/pull/4881/files

Oh yes, it is! Forgive me, as I'm still learning how to navigate all the changes in these different PRs / commits.

shonfeder commented:
@christinerose No worries at all. Thanks again for the reviews and improvements!
